Source/JavaScriptCore/inspector/InjectedScriptSource.js

```diff
@@ -223,22 +225,37 @@ let InjectedScript = class InjectedScript extends PrototypelessObjectBase
                 try {
                     resolvedArgs[i] = this._resolveCallArgument(callArgs[i]);
                 } catch (e) {
-                    return @String(e);
+                    callback(this._createThrownValue(e, objectGroupName));
+                    return;
                 }
             }
         }
 
         try {
             let func = InjectedScriptHost.evaluate("(" + expression + ")");
-            if (typeof func !== "function")
-                return "Given expression does not evaluate to a function";
-
-            return @createObjectWithoutPrototype(
-                "wasThrown", false,
-                "result", RemoteObject.create(func.@apply(object, resolvedArgs), objectGroupName, returnByValue, generatePreview),
-            );
+            if (typeof func !== "function") {
+                callback(this._createThrownValue("Given expression does not evaluate to a function", objectGroupName));
+                return;
+            }
+            let result = func.@apply(object, resolvedArgs);
+            if (awaitPromise && isDefined(result) && (InjectedScriptHost.internalConstructorName(result) === 'Promise')) {
```

Comment:
`InjectedScriptHost.internalConstructorName` seems a bit overkill.  can't we just `result instanceof @Promise`?

Style: you can drop the `(` `)`
```
if (awaitPromise && isDefined(result) && result instanceof @Promise) {
```

Reply:
ah i see what you're getting at now.  even though we do create a unique `InjectedScript` for each `JSGlobalObject` you're trying to handle the case where access to the `<iframe>` is transparent because it's the same origin (or along those lines).

in that case what about if there's a custom `class Promise { }`?  perhaps we need to create some sort of `InjectedScriptHost.isPromise(value)` that checks `value->inherits<JSPromise>()`?

if we do need something like that, id personally suggest reverting these to `instanceof @Promise` for this PR and doing a followup PR to add `InjectedScriptHost.isPromise(value)`, but im also fine having it be part of this PR if you'd prefer